examples/E2E-RAG-App/gradio_interface.py

```diff
@@ -0,0 +1,242 @@
+import asyncio
+import json
+import os
+import uuid
+from queue import Queue
+from threading import Thread
+from typing import AsyncGenerator, Generator, List, Optional
+
+import chromadb
+
+import gradio as gr
+import requests
+from chromadb.utils import embedding_functions
+from dotenv import load_dotenv
+from llama_stack_client import LlamaStackClient
+from llama_stack_client.lib.agents.agent import Agent
+from llama_stack_client.lib.agents.event_logger import EventLogger
+from llama_stack_client.types.agent_create_params import AgentConfig
+from llama_stack_client.types.memory_insert_params import Document
+
+
+# Load environment variables
+load_dotenv()
+
+HOST = os.getenv("HOST", "localhost")
+PORT = int(os.getenv("PORT", 5000))
+CHROMA_PORT = int(os.getenv("CHROMA_PORT", 6000))
+DOCS_DIR = "/root/rag_data/output"
+GRADIO_SERVER_PORT = int(os.getenv("GRADIO_SERVER_PORT", 7861))
+MODEL_NAME = os.getenv("MODEL_NAME", "Llama3.2-1B-Instruct")
+
+
+class LlamaChatInterface:
+    def __init__(self, host: str, port: int, chroma_port: int, docs_dir: str):
+        self.host = host
+        self.port = port
+        self.docs_dir = docs_dir
+        self.client = LlamaStackClient(base_url=f"http://{host}:{port}")
+        self.chroma_client = chromadb.HttpClient(host=host, port=chroma_port)
+        self.agent = None
+        self.session_id = None
+        self.memory_bank_id = "test_bank_232"
+
+    async def initialize_system(self):
+        """Initialize the entire system including memory bank and agent."""
+        await self.setup_memory_bank()
+        await self.initialize_agent()
+
+    async def setup_memory_bank(self):
+        """Set up the memory bank if it doesn't exist."""
+        providers = self.client.providers.list()
+        provider_id = providers["memory"][0].provider_id
+        collections = self.chroma_client.list_collections()
+
+        if any(col.name == self.memory_bank_id for col in collections):
+            print(f"The collection '{self.memory_bank_id}' exists.")
+        else:
+            print(
+                f"The collection '{self.memory_bank_id}' does not exist. Creating the collection..."
+            )
+            self.client.memory_banks.register(
+                memory_bank_id=self.memory_bank_id,
+                params={
+                    "embedding_model": "all-MiniLM-L6-v2",
+                    "chunk_size_in_tokens": 100,
+                    "overlap_size_in_tokens": 10,
+                },
+                provider_id=provider_id,
+            )
+            await self.load_documents()
+            print(f"Memory bank registered.")
+
+    async def load_documents(self):
```

Comment:
a little confused -- I thought the workflow was that we _pre_ ingest documents and then the chat app merely works with such a pre-ingested, already set up memory bank. because otherwise I don't understand why we needed `01_ingest_files.py`?

Reply:
let me clarify the workflow:

1. Pre-ingestion (01_ingest_files.py):
   - converts various document formats (PDF, DOCX, etc.) into standardized Markdown files
   - also run vision model on images if present in docs
   - makes documents more easily parsable and consistent
   - only needs to be run once when adding new documents

2. Memory Bank Setup (in chat app):
   - takes the pre-processed Markdown files
   - creates vector embeddings and stores them in ChromaDB
   - this is a one-time setup when initializing the chat interface, after setup, the chat app uses this existing memory bank for retrievals

So you're correct - the pre-ingestion step (01_ingest_files.py) handles document standardization and vision hanlding, while the memory bank setup handles vectorization and storage.